py/gc.c

```diff
@@ -519,7 +511,35 @@ static void gc_sweep(void) {
                         // clear finaliser flag
                         FTB_CLEAR(area, block);
                     }
-                    #endif
+                }
+                ftb >>= 1;
+                block++;
+            }
+        }
+    }
+    #endif // MICROPY_ENABLE_FINALISER
+}
+
+static void gc_sweep(void) {
+    #if MICROPY_PY_GC_COLLECT_RETVAL
+    MP_STATE_MEM(gc_collected) = 0;
+    #endif
+    // free unmarked heads and their tails
```

Comment:
This comment belongs down just after the call to `gc_sweep_run_finalisers()`.  Also probably worth moving the related variables `free_tail` and `prev_area` down as well, to the top of that for loop where they are used.

Actually, on second thought, might be worth putting the call to `gc_sweep_run_finalisers()` in `gc_collect_end()`.  Then this function (`gc_sweep`) is just doing the sweep.  Eg:
```c
void gc_collect_end(void) {               
    gc_deal_with_stack_overflow();   
    gc_sweep_run_finalisers();                   
    gc_sweep();                                         
    ...
}
```
Could even rename `gc_sweep` to `gc_sweep_reclaim_blocks` or something like that.

Reply:
Good idea. I've gone through and made those changes, and then went a step further and reordered the static functions so they read top-to-bottom in roughly the order of the GC flow, instead of the order being determined by the declaration order.

Have put this in its own commit, as it creates even more churn in the diff, but it's probably better for long-term maintainability.